### PR TITLE
Implement certificate authentication for GNUTLS

### PR DESCRIPTION
**Pull request by puiterwijk**
This implements TLS certificate authentication in the GNUTLS
implementation.
It supports mutual authentication, and determining the peer
subject DN.

Signed-off-by: Patrick Uiterwijk <patrick@puiterwijk.org>

### CODE REVIEW

File: lib/tls-gnutls.c [hunk 1]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+}
+
+static inline void
+clear_plist(gnutls_pcert_st *list, unsigned int length)
+{
+  for (unsigned int i = 0; i < length; i++)
+    gnutls_pcert_deinit(&list[i]);
+}
+
+#define NUM_CERTS_INIT 16
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: Top of file.

File: lib/tls-gnutls.c [hunk 2]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+
+  ret = gnutls_privkey_init(&k);
+  if (ret < 0)
+    goto out;
+
+  if (gnutls_url_is_supported(key_uri))
+    // PKCS11
+    ret = gnutls_privkey_import_url(k, key_uri, 0);
+  else
+  {
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: Cuddle

File: lib/tls-gnutls.c [hunk 3]
```diff
@@ -196,11 +205,65 @@ tls_write(tls_t *tls, int fd, const void *buf, size_t count)
...
+  gnutls_datum_t dn;
+
+  cert_list = gnutls_certificate_get_peers(tls->session, &cert_list_size);
+  if (cert_list_size == 0) {
+    errno = EINVAL;  // TODO
+    return -1;
+  }
+
+  ret = gnutls_x509_crt_init(&crt);
+  if (ret < 0) {
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: After this exit, I think crt leaks.  Consider `goto cleanup` or similar.

File: lib/tls-gnutls.c [hunk 4]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+
+  if (ret != 0) {
+    if (p != NULL) {
+      clear_plist(p, cert_size);
+      free(p);
+    }
+    if (k != NULL) {
+      gnutls_privkey_deinit(k);
+    }
+    return -1;
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: Shouldn't this return `ret`?

File: lib/tls-gnutls.c [hunk 5]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+  for (unsigned int i = 0; i < length; i++)
+    gnutls_pcert_deinit(&list[i]);
+}
+
+#define NUM_CERTS_INIT 16
+
+static int
+load_certificate(gnutls_session_t session, gnutls_pcert_st **pcert, unsigned int *pcert_length, gnutls_privkey_t *pkey, const char *cert_uri, const char *key_uri, const char *pin)
+{
+  gnutls_pcert_st *p;
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: This should be initialized to `NULL` for clarity.

File: lib/tls-gnutls.c [hunk 6]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+  gnutls_privkey_t k;
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: This needs to be initialized to `NULL`.

File: lib/tls-gnutls.c [hunk 7]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+  /* DNS names (only type supported) may be at most 256 bytes long */
+  char *servername;
+  size_t namelen = 256;
+  unsigned int nametype;
+
+  servername = gnutls_malloc(namelen);
+  if (servername == NULL)
+    return GNUTLS_E_MEMORY_ERROR;
+
+  for(int i = 0; ; i++) {
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: Space after `for`

File: lib/tls-gnutls.c [hunk 8]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+
+static int
+cert_clt(gnutls_session_t session, const gnutls_datum_t *req_ca_dn, int nreqs, const gnutls_pk_algorithm_t *pk_algos, int pk_algos_length, gnutls_pcert_st **pcert, unsigned int *pcert_length, gnutls_privkey_t *pkey)
+{
+  const tls_handshake_t *hs = gnutls_session_get_ptr(session);
+  char *cert_uri = NULL;
+  char *key_uri = NULL;
+  char *pin = NULL;
+  int ret = 0;
+  char **req_ca_dn_strs = gnutls_malloc(sizeof(char *) * (nreqs + 1));
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: Please move the `malloc()` call out of the variable declaration block.  (Put it next to the `NULL` check.)

File: lib/tls-gnutls.c [hunk 9]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+  gnutls_datum_t ca_dn;
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: This needs to be initialized before `goto out`.

File: lib/tls-gnutls.c [hunk 10]
```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
...
+      if (req_ca_dn_strs[i] != NULL)
+        free(req_ca_dn_strs[i]);
+    }
+    free(req_ca_dn_strs);
+  }
+  if (cert_uri != NULL) {
+    explicit_bzero(cert_uri, strlen(cert_uri));
+    free(cert_uri);
+  }
+  if (key_uri != NULL) {
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: This idiom of `if (thing != NULL) { explicit_bzero(thing); free(thing); }` gets repeated a lot.  Please make a wrapper for it and call that instead.

File: lib/tls-gnutls.c [hunk 11]
```diff
@@ -348,26 +673,100 @@ tls_handshake(tls_t *tls, int fd, bool client, const tls_handshake_t *hs)
...
+                                      tls->creds.clt.psk));
+      if (ret < 0)
+        goto error;
+    } else {
+      ret = g2e(gnutls_certificate_allocate_credentials(&tls->creds.clt.cert));
+      if (ret < 0)
+        goto error;
+
+      if (hs->clt.cert.hostname) {
+        ret = g2e(gnutls_server_name_set(tls->session,
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: Please pull the `g2e()` calls into their own lines.

File: lib/tls-gnutls.c [hunk 12]
```diff
@@ -348,26 +673,100 @@ tls_handshake(tls_t *tls, int fd, bool client, const tls_handshake_t *hs)
...
-    ret = g2e(gnutls_psk_allocate_server_credentials(&tls->creds.srv.psk));
-    if (ret < 0)
-      goto error;
-
-    gnutls_psk_set_server_credentials_function(tls->creds.srv.psk, psk_srv);
-    ret = g2e(gnutls_credentials_set(tls->session, GNUTLS_CRD_PSK,
-                                     tls->creds.srv.psk));
-    if (ret < 0)
-      goto error;
+  if (client) {
```
Comment (frozencemetery) [inline lib/tls-gnutls.c]: The control flow from here on down is hard to follow.  If there isn't a better way to write it, could you add some comments?

File: bin/main.c [hunk 13]
```diff
@@ -143,6 +143,49 @@ clt_psk_cb(void *m, char **username, uint8_t **key)
   return keydup(o, key);
 }
 
+static int
+cert_cb(void *m,
+        char **cert_uri,
+        char **key_uri,
+        char **pin)
```
Comment (npmccallum) [inline bin/main.c]: Please follow argument wrapping style in the code. This goes for all functions below.

File: bin/main.c [hunk 14]
```diff
@@ -143,6 +143,49 @@ clt_psk_cb(void *m, char **username, uint8_t **key)
...
+
+  *cert_uri = strdup(o->crtf);
+  if (!*cert_uri)
+    return -1;
+  *key_uri = strdup(o->crtk);
+  if (!key_uri)
+    return -1;
+  *pin = strdup(o->crtkp);
+  if (!pin)
+    return -1;
```
Comment (npmccallum) [inline bin/main.c]: Hrm... I know that the key URI can specify the PIN as one of its arguments. I'm not sure how I feel about having a separate PIN parameter.

File: bin/main.c [hunk 15]
```diff
@@ -160,6 +203,13 @@ on_conn(options_t *opts, int con, int in, int out, const struct addrinfo *ai)
 
       if (opts->psku)
         srv.psk = srv_psk_cb;
+      else if (opts->crtf)
+      {
```
Comment (npmccallum) [inline bin/main.c]: Don't wrap the braces here. Also, we don't do unbalanced braces. If the `else if` requires braces, the `if` gets them too.

File: bin/main.c [hunk 16]
```diff
@@ -178,6 +234,24 @@ on_conn(options_t *opts, int con, int in, int out, const struct addrinfo *ai)
...
+
+    if (opts->expectpeer != NULL) {
+      socklen_t dnlen = 256;
+      char *peer_dn = malloc(sizeof(char) * dnlen);
+      ret = getsockopt(con, IPPROTO_TLS, TLS_OPT_PEER_SUBJECT_DN, peer_dn, &dnlen);
+      if (ret != 0) {
+        fprintf(stderr, "%m: Unable to get peer subject DN!\n");
+        shutdown(con, SHUT_RDWR);
+        return STATUS_FAILURE;
+      }
```
Comment (npmccallum) [inline bin/main.c]: Please use some blank lines in here to separate the code blocks.

File: lib/tlssock.h [hunk 17]
```diff
@@ -29,14 +29,40 @@
...
 
   ssize_t (*psk)(void *misc, const char *username, uint8_t **key);
+  struct {
+    int (*getcert)(void *misc,
+                const char *servername,
+                char **cert_uri,
+                char **key_uri,
+                char **pin);
+    int client_certificate_request;  // 0 for no client cert, 1 for optional, 2 for required
+    const char *cafile;
```
Comment (npmccallum) [inline lib/tlssock.h]: If I understand this correctly, we can use the implicit value of `cafile` to determine if a client certificate can be validated. This means that `client_certificate_request` can become `bool client_certificate_required`.

File: lib/tls-gnutls.c [hunk 18]
```diff
@@ -196,11 +225,67 @@ tls_write(tls_t *tls, int fd, const void *buf, size_t count)
...
 }
 
+static int
+tls_getsockopt_peer_subject_dn(tls_t *tls, void *optval, socklen_t *optlen)
+{
+  int ret;
+  gnutls_x509_crt_t crt;
+  const gnutls_datum_t *cert_list;
+  unsigned int cert_list_size = 0;
+  gnutls_datum_t dn = {NULL, 0};
```
Comment (npmccallum) [inline lib/tls-gnutls.c]: Please sort variable definitions by line width: longest to shortest.

File: lib/tlssock.h [hunk 19]
```diff
@@ -29,14 +29,40 @@
...
 
   ssize_t (*psk)(void *misc, char **username, uint8_t **key);
+  struct {
+    int (*getcert)(void *misc,
+                const char **requested_ca_dn,
+                char **cert_uri,
+                char **key_uri,
+                char **pin);
+    const char *cafile;
+    int insecure;
```
Comment (npmccallum) [inline lib/tlssock.h]: If I understand the code correctly, `insecure` means that we trust all certificates. I wonder if we should disallow that mode altogether. In that case, `cafile = NULL` means trust the system CAs.
